### PR TITLE
backport: bitcoin#18225, #18270, #21438, #22220, #18134, #18449, #19750, #20452 (locale independence)

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
Doesn't build

File: src/rpc/util.cpp [hunk 1]
```diff
@@ -493,7 +493,7 @@ std::string RPCHelpMan::ToString() const
         if (i == 0) ret += "\nArguments:\n";
 
         // Push named argument name and description
-        sections.m_sections.emplace_back(std::to_string(i + 1) + ". " + arg.GetFirstName(), arg.ToDescriptionString());
+        sections.m_sections.emplace_back(::ToString(i + 1) + ". " + arg.GetFirstName(), arg.ToDescriptionString());
```
Comment (knst) [inline src/rpc/util.cpp:496]: looks like missing include <util/string.h>

File: src/test/specialtx_tests.cpp [hunk 2]
```diff
@@ -72,7 +72,7 @@ BOOST_AUTO_TEST_CASE(verify_mnhf_specialtx_tests)
     BOOST_CHECK(ag_sk.IsValid());
     BOOST_CHECK(ag_pk.IsValid());
 
-    uint256 verHash = uint256S(itostr(ver));
+    uint256 verHash = uint256S(ToString(ver));
```
Comment (knst) [inline src/test/specialtx_tests.cpp:75]: missing include `<util/string.h>`

File: src/init.cpp [hunk 3]
```diff
@@ -57,6 +57,7 @@
 #include <util/asmap.h>
 #include <util/error.h>
 #include <util/moneystr.h>
+#include <util/string.h>
```
Comment (knst) [inline src/init.cpp:60]: should be included `util/strencodings.h` instead of `util/string.h`

File: src/rpc/governance.cpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #include <rpc/blockchain.h>
 #include <rpc/server.h>
 #include <rpc/util.h>
+#include <util/string.h>
```
Comment (knst) [inline src/rpc/governance.cpp:21]: should be included `util/strencodings.h` instead of `util/string.h`

File: src/rpc/governance.cpp [hunk 5]
```diff
@@ -950,7 +951,7 @@ static UniValue gobject_getcurrentvotes(const JSONRPCRequest& request)
     if (!request.params[1].isNull() && !request.params[2].isNull()) {
         uint256 txid = ParseHashV(request.params[1], "Masternode Collateral hash");
         std::string strVout = request.params[2].get_str();
-        mnCollateralOutpoint = COutPoint(txid, (uint32_t)atoi(strVout));
+        mnCollateralOutpoint = COutPoint(txid, (uint32_t)LocaleIndependentAtoi<int>(strVout));
```
Comment (knst) [inline src/rpc/governance.cpp:954]: this code looks strange for me. Parse a number like a signed number and convert after that to unsigned? Is it what expected? @PastaPastaPasta

File: src/rpc/masternode.cpp [hunk 6]
```diff
@@ -16,6 +16,7 @@
 #include <rpc/server.h>
 #include <rpc/util.h>
 #include <univalue.h>
+#include <util/string.h>
```
Comment (knst) [inline src/rpc/masternode.cpp:19]: should be included `util/strencodings.h` instead of `util/string.h`

**Review by knst — CHANGES_REQUESTED** [review]
Please, check my comments.

Most of them nit/codestyle but still "request changes".

File: src/rpc/governance.cpp (hunk 5 above)
Comment (kittywhiskers) [reply src/rpc/governance.cpp:954]: `atoi` substitution was primarily scripted but will change template typedef to `uint32_t`

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -4459,20 +4457,20 @@ std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain& chain, const std::st
...
+        if (!max_fee) {
             error = AmountErrMsg("maxtxfee", gArgs.GetArg("-maxtxfee", ""));
             return nullptr;
-        }
-        if (nMaxFee > HIGH_MAX_TX_FEE) {
+        } else if (max_fee.value() > HIGH_MAX_TX_FEE) {
             warnings.push_back(_("-maxtxfee is set very high! Fees this large could be paid on a single transaction."));
         }
-        if (CFeeRate(nMaxFee, 1000) < chain.relayMinFee()) {
+        if (CFeeRate(max_fee.value(), 1000) < chain.relayMinFee()) {
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/wallet/wallet.cpp:4467]: Agreed. Change wasn't inherited from PR due to manual changes needed to resolve cherrypick conflicts, resolved.

File: src/test/util_tests.cpp [hunk 8]
```diff
@@ -1056,86 +1056,59 @@ BOOST_AUTO_TEST_CASE(util_FormatMoney)
...
 
     // Parsing negative amounts must fail
-    BOOST_CHECK(!ParseMoney("-1", ret));
+    BOOST_CHECK(!ParseMoney("-1"));
 
     // Parsing strings with embedded NUL characters should fail
-    BOOST_CHECK(!ParseMoney(std::string("\0-1", 3), ret));
-    BOOST_CHECK(!ParseMoney(STRING_WITH_EMBEDDED_NULL_CHAR, ret));
-    BOOST_CHECK(!ParseMoney(std::string("1\0", 2), ret));
+    BOOST_CHECK(!ParseMoney(std::string("\0-1", 3)));
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/test/util_tests.cpp:1109]: Changes done in PR outside the scope of backports group, was thinking about doing them but then its dependencies piled up and so I scrapped it. 

I backported only `STRING_WITH_EMBEDDED_NULL_CHAR` so I can resolve PR conflicts, I'd have preferred not having any string literal logic in this PR and saving it for later but the stars didn't align for that one.

**Review by knst — DISMISSED** [review]
LGTM now

**Review by UdjinM6 — APPROVED** [review]
some fixes were squashed into commits they do not belong to it seems but LGTM in general so utACK

**Comment by kittywhiskers** [comment]
Yeah, doing a `git rebase` with `edit` in the command line and then doing changes through Sublime Merge is known to cause odd squashing behavior, can you tell me which commits they are so I can sort them out?

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -2358,8 +2358,8 @@ static UniValue setcoinjoinamount(const JSONRPCRequest& request)
         "\nSet the goal amount in " + CURRENCY_UNIT + " for CoinJoin.\n",
         {
             {"amount", RPCArg::Type::NUM, RPCArg::Optional::NO,
-                "The default amount is " + std::to_string(DEFAULT_COINJOIN_AMOUNT) +
-                " Cannot be more than " + std::to_string(MAX_COINJOIN_AMOUNT) + " nor less than " + std::to_string(MIN_COINJOIN_AMOUNT)},
+                "The default amount is " + ToString(DEFAULT_COINJOIN_AMOUNT) +
+                " Cannot be more than " + ToString(MAX_COINJOIN_AMOUNT) + " nor less than " + ToString(MIN_COINJOIN_AMOUNT)},
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:2362]: >can you tell me which commits they are so I can sort them out?

18449: these should be in 18134 ideally but, again, not a big deal imo

File: src/wallet/rpcwallet.cpp [hunk 10]
```diff
@@ -2329,8 +2329,8 @@ static UniValue setcoinjoinrounds(const JSONRPCRequest& request)
         "\nSet the number of rounds for CoinJoin.\n",
         {
             {"rounds", RPCArg::Type::NUM, RPCArg::Optional::NO,
-                "The default number of rounds is " + std::to_string(DEFAULT_COINJOIN_ROUNDS) +
-                " Cannot be more than " + std::to_string(MAX_COINJOIN_ROUNDS) + " nor less than " + std::to_string(MIN_COINJOIN_ROUNDS)},
+                "The default number of rounds is " + ToString(DEFAULT_COINJOIN_ROUNDS) +
+                " Cannot be more than " + ToString(MAX_COINJOIN_ROUNDS) + " nor less than " + ToString(MIN_COINJOIN_ROUNDS)},
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp:2333]: same

File: test/lint/lint-locale-dependence.sh [hunk 11]
```diff
@@ -7,7 +7,6 @@ export LC_ALL=C
 
 KNOWN_VIOLATIONS=(
     "src/bench/string_cast.cpp.*atoi"
-    "src/bench/string_cast.cpp.*std::to_string"
```
Comment (UdjinM6) [inline test/lint/lint-locale-dependence.sh:10]: 19750: also -> 18134

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
re-request review once you've fixed the mangled commits (as it seems you'd like to do that), if you think the mangling isn't important to fix, please say that and re-request review.

File: test/lint/lint-locale-dependence.sh (hunk 11 above)
Comment (kittywhiskers) [reply test/lint/lint-locale-dependence.sh:10]: Resolved with f4f6c6546337e318637f05fcc4aba5ed9a64195f

File: src/wallet/rpcwallet.cpp (hunk 10 above)
Comment (kittywhiskers) [reply src/wallet/rpcwallet.cpp:2333]: Resolved with f4f6c6546337e318637f05fcc4aba5ed9a64195f

File: src/wallet/rpcwallet.cpp (hunk 9 above)
Comment (kittywhiskers) [reply src/wallet/rpcwallet.cpp:2362]: Resolve with f4f6c6546337e318637f05fcc4aba5ed9a64195f

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**